File: packages/logger/src/redact.ts
import type { LogContext } from './types.js'

/**
 * Deep-strip any object key whose name (case-insensitive) is in `keys`.
 * Arrays are walked; cycles are guarded; non-matching values pass through
 * unchanged. Returns a new structure — the input is never mutated.
 */
export function redact(value: LogContext, keys: string[]): LogContext {
  if (keys.length === 0) return value
  const lookup = new Set(keys.map((k) => k.toLowerCase()))
  return walk(value, lookup, new WeakSet()) as LogContext
}

function walk(value: unknown, keys: Set<string>, seen: WeakSet<object>): unknown {
  if (value === null || typeof value !== 'object') return value

  if (seen.has(value as object)) return '[Circular]'
  seen.add(value as object)

  if (Array.isArray(value)) {
    return value.map((item) => walk(item, keys, seen))
  }

  const out: Record<string, unknown> = {}
  for (const [k, v] of Object.entries(value as Record<string, unknown>)) {
    if (keys.has(k.toLowerCase())) {
      out[k] = '[REDACTED]'
      continue
    }
    out[k] = walk(v, keys, seen)
  }
  return out
}

File: packages/flutter/lib/src/telemetry/redact.dart
/// Deep redaction — a 1:1 port of `@refraction-ui/logger` `redact.ts`.
library;

import 'types.dart';

/// Deep-strip any object key whose name (case-insensitive) is in [keys].
/// Lists are walked; cycles are guarded; non-matching values pass through
/// unchanged. Returns a new structure — the input is never mutated.
LogContext redact(LogContext value, List<String> keys) {
  if (keys.isEmpty) return value;
  final lookup = <String>{for (final k in keys) k.toLowerCase()};
  return _walk(value, lookup, Set<Object>.identity()) as LogContext;
}

Object? _walk(Object? value, Set<String> keys, Set<Object> seen) {
  if (value == null) return value;
  if (value is String || value is num || value is bool) return value;

  if (value is Map || value is List) {
    if (seen.contains(value)) return '[Circular]';
    seen.add(value);
  } else {
    // Opaque object — not a plain map/list; pass through unchanged.
    return value;
  }

  if (value is List) {
    return value.map((item) => _walk(item, keys, seen)).toList();
  }

  final map = value as Map;
  final out = <String, Object?>{};
  map.forEach((k, v) {
    final key = k.toString();
    if (keys.contains(key.toLowerCase())) {
      out[key] = '[REDACTED]';
    } else {
      out[key] = _walk(v, keys, seen);
    }
  });
  return out;
}
